File: C#/DiffProcessor/SortedLimitedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DiffProcessor
{
	public class SortedLimitedList<T> where T : IComparable<T>
	{
		public class Entry
		{
			public T Value { get; internal set; }
			public Entry Previous { get; internal set; }
			public Entry Next { get; internal set; }
			public SortedLimitedList<T> List { get; internal set; }
		}

		protected Entry freePool;
		protected Entry first;
		protected Entry last;

		protected Entry allocate()
		{
			if (freePool == null)
				throw new OutOfMemoryException();
			Entry result = freePool;
			freePool = freePool.Next;

			return result;
		}

		protected void free(Entry entry)
		{
			entry.Next = freePool;
			freePool = entry;
		}

		protected void check(Entry entry)
		{
			Entry before = entry.Previous;
			if (before != null && before.Value.CompareTo(entry.Value) > 0)
				throw new Exception("List not sorted");

			Entry after = entry.Next;
			if (after != null && after.Value.CompareTo(entry.Value) < 0)
				throw new Exception("List not sorted");
		}

		public SortedLimitedList(Int32 limit)
		{
			first = null;
			last = null;
			freePool = null;
			Limit = limit;
			PerformedOperations = 0;
			Count = 0;
			for (Int32 i = 0; i < limit; ++i)
			{
				Entry current = new Entry();
				current.List = this;
				free(current);
			}
		}

		public void FromArray(T[] array)
		{
			Clear();
			for (Int32 i = 0; i < array.Length; ++i)
				AddLast(array[i]);
			PerformedOperations = 0;
		}

		public void FromList(List<T> array)
		{
			Clear();
			for (Int32 i = 0; i < array.Count; ++i)
				AddLast(array[i]);
			PerformedOperations = 0;
		}

		public override String ToString()
		{
			StringBuilder builder = new StringBuilder();
			Entry e = First;
			while (e != null)
			{
				builder.Append(e.Value);
				if (e.Next != null)
				{
					builder.Append(", ");
				}
				e = e.Next;
			}

			return builder.ToString();
		}

		public void AddAfter(Entry after, T value)
		{
			if (after == null)
				AddLast(value);
			else
			{
				Entry entry = allocate();

				entry.Value = value;
				entry.Previous = after;
				entry.Next = after.Next;
				++Count;

				if (after.Next != null)
					after.Next.Previous = entry;
				else
					last = entry;

				after.Next = entry;
				++PerformedOperations;
				check(entry);
			}
		}

		public void AddBefore(Entry before, T value)
		{
			if (before == null)
				AddFirst(value);
			else
			{
				Entry entry = allocate();

				entry.Value = value;
				entry.Previous = before.Previous;
				entry.Next = before;
				++Count;

				if (before.Previous != null)
					before.Previous.Next = entry;
				else
					first = entry;

				before.Previous = entry;
				++PerformedOperations;
				check(entry);
			}
		}

		public void AddLast(T value)
		{
			Entry entry = allocate();

			entry.Value = value;
			entry.Previous = last;
			entry.Next = null;
			++Count;

			if (last != null)
				last.Next = entry;
			else
				first = entry;

			last = entry;
			++PerformedOperations;
			check(entry);
		}

		public void AddFirst(T value)
		{
			Entry entry = allocate();

			entry.Value = value;
			entry.Previous = null;
			entry.Next = first;
			++Count;

			if (first != null)
				first.Previous = entry;
			else
				last = entry;

			first = entry;
			++PerformedOperations;
			check(entry);
		}

		public void Remove(Entry entry)
		{
			if (entry.Previous != null)
				entry.Previous.Next = entry.Next;
			else
				first = entry.Next;

			if (entry.Next != null)
				entry.Next.Previous = entry.Previous;
			else
				last = entry.Previous;

			++PerformedOperations;
			--Count;
			free(entry);
		}

		public void Clear()
		{
			Entry entry = First, next;
			while (entry != null) 
			{
				next = entry.Next;
				Remove(entry);
				entry = next;
			}
			PerformedOperations = 0;
			Count = 0;
		}

		public Int32 Count { get; private set; }
		public Int32 Limit { get; private set; }
		public Int32 PerformedOperations { get; private set; }

		public Boolean Equals(SortedLimitedList<T> list)
		{
			Entry e1 = First;
			Entry e2 = list.First;
			while (e1 != null && e2 != null)
			{
				if (e1.Value.CompareTo(e2.Value) != 0)
					return false;
				e1 = e1.Next;
				e2 = e2.Next;
			}

			return e1 == null && e2 == null;
		}

		public Entry First { get {return first; } }
		public Entry Last { get { return last; } }
	}
}

File: Java/src/diffprocessor/SortedLimitedList.java
package diffprocessor;
 
import sun.plugin.dom.exception.InvalidStateException;
 
import java.util.List;
 
/**
 * Created by VavilauA on 6/19/2015.
 */
public class SortedLimitedList<T extends Comparable<T>> {
 
    public static class Entry<S extends Comparable<S>> {
        private S value;
        private Entry<S> previous;
        private Entry<S> next;
        private SortedLimitedList<S> list;
 
        public S getValue() {
            return value;
        }
 
        public Entry<S> getPrevious() {
            return previous;
        }
 
        public Entry<S> getNext() {
            return next;
        }
 
        public SortedLimitedList<S> getList() {
            return list;
        }
    }
 
    Entry<T> freePool;
    Entry<T> first;
    Entry<T> last;
    int count;
    int limit;
    int performedOperations;
 
    private Entry<T> allocate() {
        if (freePool == null)
            throw new OutOfMemoryError();
        Entry<T> result = freePool;
        freePool = freePool.next;
 
        return result;
    }
 
    protected void free(Entry<T> entry) {
        entry.next = freePool;
        freePool = entry;
    }
 
    protected void check(Entry<T> entry) {
        Entry<T> before = entry.previous;
        if (before != null && before.value.compareTo(entry.value) > 0)
            throw new InvalidStateException("List not sorted");
 
        Entry<T> after = entry.next;
        if (after != null && after.value.compareTo(entry.value) < 0)
            throw new InvalidStateException("List not sorted");
    }
 
    public SortedLimitedList(int limit) {
        first = null;
        last = null;
        freePool = null;
        this.limit = limit;
        this.performedOperations = 0;
        this.count = 0;
        for (int i = 0; i < limit; ++i) {
            Entry current = new Entry();
            current.list = this;
            free(current);
        }
    }
 
    public void fromArray(T[] array) {
        clear();
        for (int i = 0; i < array.length; ++i)
            addLast(array[i]);
        performedOperations = 0;
    }
 
    public void fromList(List<T> array) {
        clear();
        for (int i = 0; i < array.size(); ++i)
            addLast(array.get(i));
        performedOperations = 0;
    }
 
    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        Entry e = first;
        while (e != null)
        {
            builder.append(e.value);
            if (e.next != null)
            {
                builder.append(", ");
            }
            e = e.next;
        }
 
        return builder.toString();
    }
 
    public void addAfter(Entry<T> after, T value) {
        if (after == null)
            addLast(value);
        else
        {
            Entry<T> entry = allocate();
 
            entry.value = value;
            entry.previous = after;
            entry.next = after.next;
            ++count;
 
            if (after.next != null)
                after.next.previous = entry;
            else
                last = entry;
 
            after.next = entry;
            ++performedOperations;
            check(entry);
        }
    }
 
    public void addBefore(Entry<T> before, T value) {
        if (before == null)
            addFirst(value);
        else
        {
            Entry<T> entry = allocate();
 
            entry.value = value;
            entry.previous = before.previous;
            entry.next = before;
            ++count;
 
            if (before.previous != null)
                before.previous.next = entry;
            else
                first = entry;
 
            before.previous = entry;
            ++performedOperations;
            check(entry);
        }
    }
 
    public void addLast(T value) {
        Entry<T> entry = allocate();
 
        entry.value = value;
        entry.previous = last;
        entry.next = null;
        ++count;
 
        if (last != null)
            last.next = entry;
        else
            first = entry;
 
        last = entry;
        ++performedOperations;
        check(entry);
    }
 
    public void addFirst(T value) {
        Entry<T> entry = allocate();
 
        entry.value = value;
        entry.previous = null;
        entry.next = first;
        ++count;
 
        if (first != null)
            first.previous = entry;
        else
            last = entry;
 
        first = entry;
        ++performedOperations;
        check(entry);
    }
 
    public void remove(Entry<T> entry) {
        if (entry.previous != null)
            entry.previous.next = entry.next;
        else
            first = entry.next;
 
        if (entry.next != null)
            entry.next.previous = entry.previous;
        else
            last = entry.previous;
 
        ++performedOperations;
        --count;
        free(entry);
    }
 
    public void clear() {
        Entry<T> entry = first, next;
        while (entry != null)
        {
            next = entry.next;
            remove(entry);
            entry = next;
        }
        performedOperations = 0;
    }
 
    public boolean equals(SortedLimitedList<T> list) {
        Entry<T> e1 = first;
        Entry<T> e2 = list.first;
        while (e1 != null && e2 != null)
        {
            if (e1.value.compareTo(e2.value) != 0)
                return false;
            e1 = e1.next;
            e2 = e2.next;
        }
 
        return e1 == null && e2 == null;
    }
 
    public Entry<T> getFirst() {
        return first;
    }
 
    public Entry<T> getLast() {
        return last;
    }
 
    public int getCount() {
        return count;
    }
 
    public int getLimit() {
        return limit;
    }
 
    public int getPerformedOperations() {
        return performedOperations;
    }
}